web/internal/metadata.bzl

```diff
@@ -0,0 +1,186 @@
+  """Generates a web_test metadata file with specified contents.
+
+  Args:
+    ctx: a Skylark rule context.
+    output: File object. The file to write the metadata to.
+    environment: string. The Web Test Launcher environment name.
+    browser_label: Label. The label for a browser rule.
+    test_label: Label. The label for the test being executed.
+    web_test_files: sequence of web_test_file structs. The named files needed
+      for this configuration.
+  """
+  content = '{\n  "_comment": "generated file for %s"' % ctx.label
+  if capabilities:
+    content += ',\n  "capabilities": ' + capabilities
+  if environment:
+    content += ',\n  "environment": "' + environment + '"'
+  if browser_label:
+    content += ',\n  "browserLabel": "%s"' % browser_label
+  if test_label:
+    content += ',\n  "testLabel": "%s"' % test_label
+  if web_test_files:
+    content += ',\n  "webTestFiles": ['
+    first = True
+    for f in web_test_files:
+      if first:
+        first = False
+      else:
+        content += ','
+      content += '\n' + _web_test_files_to_string(ctx, f, '    ')
+    content += '\n  ]'
+  content += '\n}\n'
+
+  ctx.file_action(output=output, content=content, executable=False)
+
+
+def _web_test_files_to_string(ctx, files_struct, indent=''):
+  """Converts a web_test_files struct to a JSON string.
+
+  Args:
+    ctx: a Skylark rule context.
+    files_struct: a web_test_files struct.
+    indent: string. the base indentation that should be used when writing
+      generating the JSON-encoded string.
+  Returns:
+    a JSON-encoded string representing the web_test_files object.
+  """
+  result = indent + '{\n'
+  if files_struct.archive_file:
+    result += indent + '  "archiveFile": "' + files.long_path(
+        ctx, files_struct.archive_file) + '",\n'
+  result += indent + '  "namedFiles": {'
+  first = True
+  for n, f in files_struct.named_files.items():
+    if first:
+      first = False
+    else:
+      result += ','
+    result += '\n' + indent + '    "' + n + '": "'
+    if type(f) == type(''):
+      result += f
+    else:
+      result += files.long_path(ctx, f)
+    result += '"'
+  result += '\n' + indent + '  }\n' + indent + '}'
+  return result
+
+
+def _get_files(ctx, attr_names):
+  """Finds all of the web_test_metadata files in given attributes.
+
+  It ensures that all web_test_metadata files position in the returned list is
+  based on the last attribute in attr_names that they appear in.
+
+  Args:
+    ctx: a Skylark rule context.
+    attr_names: sequence of strings. The names of attributes to get 
+      web_test_metadata files from. Attributes earlier in the list will
+      return files earlier in the result.
+  Returns:
+    A sequence of all of the web_test_metadata files in the specificied 
+    attributes.
+  """
+  metadata_files = set()
+
+  for attr_name in reversed(attr_names):
+    attr = getattr(ctx.attr, attr_name)
+    if lists.is_list_like(attr):
```

Comment:
There's a lot of clever reflection going on in these methods that might not be necessary.

While working on cl/137763822 with @tituswinters a few days ago, he shared with me go/totw/60 which argues why we should reconsider writing clever code. It's a worthwhile read, but another thing to consider is this: I understand how to use the Skylark APIs, but I don't understand how to follow the behavior of this framework built on top of it.

Maybe in a follow up PR, you'll consider refactoring the code to be more unfancy and less frameworky?